### PR TITLE
GCLOUD- pipeline generator and build pipeline for node

### LINKED ISSUE #194: Create a Pipeline Generator for Google Cloud
REQUIREMENTS

- [ ] Be compliant with Hangar way of working.

### LINKED ISSUE #259: Build Pipeline Definition and Assets for Node projects on Google Cloud

### PR DESCRIPTION
**Pull request by ultymatom**
Closes #194
Closes #259

### CODE REVIEW

File: scripts/pipelines/gcloud/pipeline_generator.sh [hunk 1]
```diff
@@ -0,0 +1,154 @@
+#!/bin/bash
+set -e
+FLAGS=$(getopt -a --options c:n:d:a:b:l:i:u:p:h --long "config-file:,pipeline-name:,local-directory:,artifact-path:,target-branch:,language:,build-pipeline-name:,sonar-url:,sonar-token:,image-name:,registry-user:,registry-password:,resource-group:,storage-account:,storage-container:,cluster-name:,s3-bucket:,s3-key-path:,quality-pipeline-name:,dockerfile:,test-pipeline-name:,aws-access-key:,aws-secret-access-key:,aws-region:,ci-pipeline-name:,help" -- "$@")
+
+eval set -- "$FLAGS"
+while true; do
+    case "$1" in
+        -c | --config-file)       configFile=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/gcloud/pipeline_generator.sh:8]: @sonatype-lift ignore

We call it in the lib file

File: scripts/pipelines/gcloud/pipeline_generator.sh [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    hangarPath=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && cd ../../.. && pwd )
+}
+
+# Function that adds the variables to be used in the pipeline.
+function addCommonPipelineVariables {
+    if test -z "${artifactPath}"
+    then
+        echo "Skipping creation of the variable artifactPath as the flag has not been used."
+    else
+        [[ "$subsitutionVariable" == "" ]] && artifactPathSubStr="_ARTIFACT_PATH=${artifactPath}" || artifactPathSubStr=",_ARTIFACT_PATH=${artifactPath}"
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/gcloud/pipeline_generator.sh:62]: @sonatype-lift ignore
variable set in the cnfig file

File: scripts/pipelines/gcloud/templates/build/build-pipeline.yml.template [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  args: ['sh', '.pipelines/scripts/build.sh']
+
+- name: ${_DOCKER_IMAGE_STEPS}
+  args: ['tar', '-cf', 'build_${BUILD_ID}.tar', '${_TARGET_DIRECTORY}']
+
+# mark to insert step for additonal artifact #
+
+artifacts:
+  objects:
+    location: gs://${PROJECT_ID}_cloudbuild/pipelinesArtifacts/${pipelineName}/end_time_$(date +%Y-%m-%d_%Hh%M)_commit_${SHORT_SHA}
```
Comment (ultymatom) [inline scripts/pipelines/gcloud/templates/build/build-pipeline.yml.template:12]: I just have an issue with the date, the timezone is an issue, should we choose one by default ? or use the one from the local of the user ?

File: scripts/pipelines/gcloud/pipeline_generator.sh [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        git branch -D "$sourceBranch" ; git push origin --delete "$sourceBranch"
+    fi
+}
+
+function createTrigger {
+    gitOriginUrl=$(git config --get remote.origin.url)
+    gCloudProject=$(echo "$gitOriginUrl" | cut -d'/' -f5)
+    gCloudRepo=$(echo "$gitOriginUrl" | cut -d'/' -f7)
+    #debug
+    gcloud beta builds triggers create cloud-source-repositories --repo="$gCloudRepo" --branch-pattern="$branchTrigger"  --build-config="${pipelinePath}/${yamlFile}" --project="$gCloudProject" --name="$pipelineName" --description="$triggerDescription" --substitutions "${subsitutionVariable}${artifactPathSubStr}"
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/gcloud/pipeline_generator.sh:90]: @sonatype-lift ignore

File: scripts/pipelines/gcloud/templates/build/build-pipeline.yml.template (hunk 3 above)
Comment (ultymatom) [reply scripts/pipelines/gcloud/templates/build/build-pipeline.yml.template:12]: knowing that the handling of timezone is different between git bash and a regular bash

File: scripts/pipelines/gcloud/pipeline_generator.sh [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+obtainHangarPath
+
+# Load common functions
+. "$hangarPath/scripts/pipelines/common/pipeline_generator.lib"
+
+if [[ "$help" == "true" ]]; then help; fi
```
Comment (mcerverc) [inline scripts/pipelines/gcloud/pipeline_generator.sh:98]:
```suggestion
if [[ "$help" == "true" ]]; then help; exit; fi
```

**Comment by mcerverc** [comment]
Better use -n "$varname" instead of ! -z "$varname" to check if a variable is filled.
Here (I cannot suggest changes as this code is not from this commit):
https://github.com/devonfw/hangar/blob/16be53b08b8376bc8f90672cb3ca3ac0a87596a8/scripts/pipelines/common/pipeline_generator.lib#L176

File: scripts/pipelines/gcloud/pipeline_generator.sh (hunk 5 above)
Comment (mcerverc) [reply scripts/pipelines/gcloud/pipeline_generator.sh:98]: Also here(not from this commit) :
https://github.com/ultymatom/hangar/blob/16be53b08b8376bc8f90672cb3ca3ac0a87596a8/scripts/pipelines/github/pipeline_generator.sh#L128
and here:
https://github.com/ultymatom/hangar/blob/16be53b08b8376bc8f90672cb3ca3ac0a87596a8/scripts/pipelines/azure-devops/pipeline_generator.sh#L145

Comment (ultymatom) [reply scripts/pipelines/gcloud/pipeline_generator.sh:98]: the exit command is already inside the help function

**Review by mcerverc — APPROVED** [review]

File: scripts/pipelines/gcloud/pipeline_generator.sh [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+}
+
+function merge_branch {
+    # Check if a target branch is supplied.
+    if test -z "$targetBranch"
+    then
+        # No branch specified in the parameters, no Pull Request is created, the code will be stored in the current branch.
+        echo -e "${green}No branch specified to do the merge, changes left in the ${sourceBranch} branch."
+        exit
```
Comment (jrodenasCapgemini) [inline scripts/pipelines/gcloud/pipeline_generator.sh:73]:
```suggestion
        echo -e "${green}No branch specified to do the merge, changes left in the ${sourceBranch} branch."
        echo -ne "${white}"
        exit
```

File: documentation/src/common_templates/setup-build-pipeline.asciidoc [hunk 7]
```diff
@@ -28,7 +28,8 @@ pipeline_generator.sh \
   -d <project local path> \
   [-t <target-directory>] \
   [-b <branch>] \
-  [-w]
+ifndef::no-PR-or-MR[  [-w]]
+
```
Comment (albertdb) [inline documentation/src/common_templates/setup-build-pipeline.asciidoc:32]:
```suggestion
```

File: scripts/pipelines/gcloud/pipeline_generator.sh [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+function createTrigger {
+    gitOriginUrl=$(git config --get remote.origin.url)
+    gCloudProject=$(echo "$gitOriginUrl" | cut -d'/' -f5)
+    gCloudRepo=$(echo "$gitOriginUrl" | cut -d'/' -f7)
+    # We check if the bucket we needed exists, we create it if not
+    if (gcloud storage ls --project="${gCloudProject}" | grep "${gCloudProject}_cloudbuild" >> /dev/null)
+    then
+      echo -e "${green}Bucket ${gCloudProject}_cloudbuild already existing.${white}"
```
Comment (albertdb) [inline scripts/pipelines/gcloud/pipeline_generator.sh:91]:
```suggestion
      echo -e "${green}Bucket ${gCloudProject}_cloudbuild already exists.${white}"
```

**Review by albertdb — CHANGES_REQUESTED** [review]

**Review by albertdb — APPROVED** [review]

**Pull request merged**